### PR TITLE
Fix race condition with working set not reflecting reality during integrity check.

### LINKED ISSUE #423: VersionConflict with globally installed packages
The following relates to buildout 2.9.5

If there is a package installed globally, eg. zope.interface 4.1.1  in /usr/lib/python2.7/dist-packages, and buildout tries to install a lower version, eg. zope.interface==3.7.0, then buildout fails with a VersionConflict.  Re-running buildout directly afterwards succeeds, as the 3.7.0 egg is installed before failure.

    File "/home/paul/git/grokproject/eggs/zc.buildout-2.9.5-py2.7.egg/zc/buildout/easy_install.py", line 714, in install
      for dist in self._get_dist(req, ws):
    File "/home/paul/git/grokproject/eggs/zc.buildout-2.9.5-py2.7.egg/zc/buildout/easy_install.py", line 570, in _get_dist
      dist = self._env.best_match(requirement, ws)
    File "build/bdist.linux-x86_64/egg/pkg_resources/__init__.py", line 1111, in best_match
    dist = working_set.find(req)
    File "build/bdist.linux-x86_64/egg/pkg_resources/__init__.py", line 715, in find
    raise VersionConflict(dist, req)

Virtualenv does not help in resolving this, even with the --no-site-packages.

Buildout no longer has a -s flag to ignore site-packages.

I would expect buildout to ignore the conflict and continue, in fact not report a conflict at all, as the potential conflict is directly resolved by the action of installing the egg.

### PR DESCRIPTION
**Pull request by prsephton**
In easy_install.Installer._get_dist, a package is retrieved and installed in the eggs folder (eg. ~/.buildout/eggs).

Previously, the program would re-scan the environment for changes, check that the newly downloaded package was accessible and return a list of packages (allowing for dependencies) which were installed.  These packages would then be added to the working set.

The problem was that the accessibility check was done against a working set that did not yet reflect the newly downloaded packages.  This was not generally a problem, except where a conflicting package was already installed (for example in sdist).  In such a case, buildout would fail, reporting a version conflict, but running buildout directly afterward would succeed because the environment and working set would contain the packages installed in the eggs folder, and there would be no conflict.

This patch moves the "ws.add(dist)" calls from where they previously were, to inside the _get_dist() method, and adds the "replace=True" flag which effectively replaces any conflicting working set packages with the newly downloaded ones.  Now when we do:

            self._env_rescan_dest()
            dist = self._env.best_match(requirement, ws)

we are testing against a working set that reflects the true file system state, and buildout runs consistently.

Fixes #423

### CODE REVIEW

**Comment by reinout** [comment]
Your description of the problem is not totally clear to me. At least, I'm a bit confused by your terminology:

- An "sdist" is a source distribution, a .tgz you download from pypi. It is a package. So if you say "except where a conflicting package was already installed (for example in sdist)", it doesn't make much sense.

- A "working set" is an in-memory list of packages that buildout explicity adds packages to. So it isn't the case that everything in the `eggs/` directory is automatically added to it. So it isn't automatically in sync when you run buildout again. 

- File system state: what's in `eggs/` isn't automatically what's in the working set. You can set the eggs dir to `~/buildout-cache/eggs/`, for instance, sharing/caching eggs between many different buildouts.

Can you clarify a bit?

**Comment by prsephton** [comment]
Hi, @reinout, 

> An "sdist" is a source distribution, a .tgz you download from pypi

Sorry, I misspoke.  What I meant to say, is "excepting where an existing package was installed, for example as an installed system package".  

In my case for example, my Debian distro has an unrelated app which depends on zope.interface, and this gets installed as a system package.  My buildout-installed zope.interface has a pinned version which pre-dates the one Debian installs.

For the sake of discussion, lets stick to zope.interface as an example.

> A "working set" is an in-memory list of packages that buildout explicity adds packages to.

This is true, but `pkg_resources.WorkingSet.__init__()` (line 608) initialises itself with:

        if entries is None:
            entries = sys.path

        for entry in entries:
            self.add_entry(entry)

WorkingSet's `__iter__()` comment reads as follows:

    def __iter__(self):
        """Yield distributions for non-duplicate projects in the working set

        The yield order is the order in which the items' path entries were
        added to the working set.
        """

`pkg_resources._initialize_master_working_set()` also explicitly adds all known packages to the working set on startup, one at a time.

Therefore,  WorkingSet already knows about the zope.interface (and all other packages) installed at a system level.

On the first run, buildout downloads our pinned version of zope.interface and installs it in the eggs dir.  At line 569 of existing code for `easy_install.Installer._get_dist()`, it then calls:

            self._env_rescan_dest()
            dist = self._env.best_match(requirement, ws)

which reports a version conflict, as the newly downloaded and installed zope.interface conflicts with the already known system package.  At this point, our new egg is not known to us in `Installer._env` or working set, although it has been installed in the eggs dir.  We currently add the egg to the working set only *after* `_get_dist()` returns.

> File system state: what's in eggs/ isn't automatically what's in the working set.

On the second run, when installing zope,interface, easy_install.install() creates an easy_install.Installer() with the eggs dir as dest, and `easy_install.__init__()` creates a `self._env` that contains all of the known eggs in the dest.  In `Installer._get_dist()` (line 532), `Installer._satisfied()` returns the dist from the eggs dir (self._env), and there is no version conflict.  When `_get_dist()` returns, the dist is added to the working set.  

So by the time all is done, the working set contains the *relevant* eggs from the eggs dir.

So in conclusion, there is inconsistent behavior for buildout between the first and subsequent runs if a pinned package version conflicts with an already installed system package.

To get buildout to succeed on the first try, after downloading and installing the package at easy_install.Installer._get_dist(), line 557, we need to first add the package to the working set, and *then* do `dist = self._env.best_match(requirement, ws)` which will no longer report a version conflict (unless there really is one).

I hope this helps.

**Comment by reinout** [comment]
I don't totally understand the full implications of the change yet, but it *smells* like a good one. If I'm right, it might especially solve some setuptools-from-the-main-sitepackages issues. Which is a very good thing :-)

There are some errors in the tests. All of them seem related to the path order of setuptools. @prsephton, can you check whether those changes in the test output are what you're expecting? (And if they are: change the tests to reflect the new situation).

**Comment by prsephton** [comment]
Apologies for taking a while over this update, but I have been doing, redoing, exploring and going over things with a fine toothcomb, and reducing the footprint of the change as far as possible.

@leorochael Thanks for your feedback. For now, just to avoid confusion I would like to make as few changes as possible.  If everyone is happy eventually, perhaps we can do the rename of _get_dist() to _get_dist_and_update_working_set() (or whatever).  Since _maybe_add_setuptools() calls _get_dist() it would probably not be a good idea to inline this function within _get_dist().

@reinout The reversal of package lists inside of working set all have to do with the use of the "replace=True" argument to ws.add().  By first checking for the existence of a package in the ws prior to calling ws.add(), most of those additions which would have affected the ordering are eliminated, and this minimised the number of tests which broke.  Indeed, the PR without the "replace=True" argument, all of the original tests succeed without any changes to the test cases at all.

As for the implications of this change, I think it addresses a very specific behaviour, where the working set may already contain a package which we are trying to install as an egg, and the new egg conflicts with the system package. Other than that edge case, I believe behaviour is identical, or near as dammit to existing implementation.

Regards

**Comment by leorochael** [comment]
> @leorochael Thanks for your feedback. For now, just to avoid confusion I would like to make as few changes as possible.  If everyone is happy eventually, perhaps we can do the rename of _get_dist() to _get_dist_and_update_working_set() (or whatever).  Since _maybe_add_setuptools() calls _get_dist() it would probably not be a good idea to inline this function within _get_dist().

I'm fine with `_get_dist_and_update_working_set()`.

The idea of inlining `_maybe_add_setuptools()` would be to also change it to not call `_get_dist()` recursively but to detect whether adding setuptools is necessary as soon as possible so as to interpret the parameters as not only requesting the original dist, but setuptools as well.

That said, I'm fine leaving the inlining change for later.

> As for the implications of this change, I think it addresses a very specific behaviour, where the working set may already contain a package which we are trying to install as an egg, and the new egg conflicts with the system package. Other than that edge case, I believe behaviour is identical, or near as dammit to existing implementation.

Seems fine to me. I would only ask that you collapse all your commits into the first.

@jimfulton, any opinion?

**Comment by reinout** [comment]
I'm getting the impression that this will really help with system packages that conflict with versions we've specified ourselves.
I must say I'm not 100% sure that there won't be any side effects. But... as there were only minimal changes in the tests, I assume we're safe here.
So... I'm merging it and releasing it. Then we'll know for sure :-)

**Comment by reinout** [comment]
@prsephton, I've released 2.11.0: can you double-check if it fixes your issue?
Thanks for the pull request!

**Comment by prsephton** [comment]
Thanks, @reinout , its working quite sweetly!  Thanks for the review and release.
